File: problems/366-find-leaves-of-binary-tree/dfs-post-order.cpp
/**
 * 后序遍历
 * 
 * 时间：`O(N)`, 4ms
 */

class Solution {
public:
    vector<vector<int>> findLeaves(TreeNode* root) {
        dfs(root);
        return res;
    }

    int dfs(TreeNode* node) {
        if (!node) return -1;

        // 左
        int left = dfs(node->left);
        // 右
        int right = dfs(node->right);
        // 本结点
        int curr = max(left, right) + 1;
        if (curr >= res.size()) res.push_back({});
        res[curr].push_back(node->val);
        return curr;
    }

private:
    vector<vector<int>> res;
};

File: 剑指offer/27/dfs-post-order.js
/**
 * 后序遍历
 */
var mirrorTree = function (root) {
  if (!root) return root;

  // 先递归子树
  root.left = mirrorTree(root.left);
  root.right = mirrorTree(root.right);
  // 再反转本结点
  [root.left, root.right] = [root.right, root.left];
  return root;
};